### PR TITLE
Feature 48 - Bucket expiration support added

### PR DESCRIPTION
**Pull request by kirillston**

### CODE REVIEW

**Comment by kirillston** [comment]
Corresponding changes on tile are presented in following PR:
https://github.com/thecodeteam/ecs-broker-tile/pull/36

File: src/test/java/com/emc/ecs/servicebroker/service/EcsServiceTest.java [hunk 1]
```diff
@@ -782,6 +815,197 @@ public void changeBucketPlanTestDisableMetadata() throws Exception {
...
+     *
+     * @throws Exception when mocking fails
+     */
+    @Test
+    public void changeBucketExpirationTestNoRules() throws Exception {
+        // no Policy -> grantPolicy & update w null rules
+//        setupGetBucketPolicyTest(null, BUCKET_NAME);
+//        setupUpdateBucketPolicyTest();
+//        setupGetExpirationTest(0, 0, PREFIX + BUCKET_NAME);
+//        setupUpdateExpirationTest(THIRTY);
```
Comment (BassinD) [inline src/test/java/com/emc/ecs/servicebroker/service/EcsServiceTest.java:830]: Commented code

Comment (kirillston) [reply src/test/java/com/emc/ecs/servicebroker/service/EcsServiceTest.java]: Removed unnecessary dev-time comments

**Review by BassinD — APPROVED** [review]

File: src/main/java/com/emc/ecs/servicebroker/service/EcsService.java [hunk 2]
```diff
@@ -876,6 +920,66 @@ public String getDefaultNamespace() {
...
+            logger.info("Applying bucket expiration on '{}': {} days", bucketName, days);
+            BucketExpirationAction.update(broker, prefix(bucketName), days, null);
+        } else {
+            List<LifecycleRule> rules = new ArrayList<>(configuration.getRules());
+            for (LifecycleRule rule : rules) {
+                if (rule.getStatus() == LifecycleRule.Status.Enabled && rule.getId().startsWith(BucketExpirationAction.RULE_PREFIX)) {
+                    if (rule.getExpirationDays() != days) {
+                        logger.info("Changing bucket expiration on '{}': {} days instead of {} days", bucketName, days, rule.getExpirationDays());
+                        rules.remove(rule);
+                        BucketExpirationAction.update(broker, prefix(bucketName), days, rules);
```
Comment (rogalv) [inline src/main/java/com/emc/ecs/servicebroker/service/EcsService.java:955]: Namespace parameter is not used here, is it intentional? What if bucket is not in default namespace?

File: src/main/java/com/emc/ecs/servicebroker/service/EcsService.java [hunk 3]
```diff
@@ -876,6 +920,66 @@ public String getDefaultNamespace() {
...
+
+        LifecycleConfiguration configuration = BucketExpirationAction.get(broker, prefix(bucketName));
+
+        if (configuration != null && configuration.getRules() != null) {
+            List<LifecycleRule> rules = new ArrayList<>(configuration.getRules());
+            for (LifecycleRule rule : rules) {
+                if (rule.getStatus() == LifecycleRule.Status.Enabled && rule.getId().startsWith(BucketExpirationAction.RULE_PREFIX)) {
+                    logger.info("Removing bucket expiration rule on bucket '{}' in '{}' ({} days)", prefix(bucketName), namespace, rule.getExpirationDays());
+                    rules.remove(rule);
+                    BucketExpirationAction.delete(broker, prefix(bucketName), rule.getId(), rules);
```
Comment (rogalv) [inline src/main/java/com/emc/ecs/servicebroker/service/EcsService.java:976]: Same here - namespace parameter is not used.

File: src/main/java/com/emc/ecs/servicebroker/service/s3/BucketExpirationAction.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        S3Config s3Config = new S3Config(new URI(broker.getRepositoryEndpoint()))
+                .withIdentity(broker.getPrefixedUserName())
+                .withSecretKey(broker.getRepositorySecret());
+        S3Client s3Client = new S3JerseyClient(s3Config);
+
+        LifecycleConfiguration lifecycle;
+        try {
+            lifecycle = s3Client.getBucketLifecycle(bucketName);
+        } catch (S3Exception exception) {
+            logger.debug("Object user '{}' does not have Lifecycle Management policy on bucket '{}'", broker.getPrefixedUserName(), bucketName);
```
Comment (rogalv) [inline src/main/java/com/emc/ecs/servicebroker/service/s3/BucketExpirationAction.java:53]: Is it a debug level message? User should already have needed access at this point, so we should issue a warning.

Comment (kirillston) [reply src/main/java/com/emc/ecs/servicebroker/service/s3/BucketExpirationAction.java]: Changed logging level

File: src/main/java/com/emc/ecs/servicebroker/service/EcsService.java (hunk 2 above)
Comment (kirillston) [reply src/main/java/com/emc/ecs/servicebroker/service/EcsService.java]: Added namespace in expiration action requests

File: src/main/java/com/emc/ecs/servicebroker/service/EcsService.java (hunk 3 above)
Comment (kirillston) [reply src/main/java/com/emc/ecs/servicebroker/service/EcsService.java]: Added namespace in expiration action requests

File: src/test/java/com/emc/ecs/common/Fixtures.java [hunk 5]
```diff
@@ -594,4 +598,38 @@ public static DeleteServiceInstanceBindingRequest bucketBindingRemoveFixture() {
...
+        if (expirationDays != 0 & rulesNumber == 1) {
+            return new LifecycleConfiguration().withRules(expirationRule);
+        }
+
+        int nonExpirationRules = expirationDays == 0 ? rulesNumber : rulesNumber - 1;
+
+        LifecycleRule[] rules = new LifecycleRule[rulesNumber];
+        for (int i = 0; i < nonExpirationRules; i++) {
+            rules[i] = new LifecycleRule(NON_LC_RULE_PREFIX + UUID.randomUUID().toString(), i + bucket,
+                    i % 2 == 0 ? LifecycleRule.Status.Disabled : LifecycleRule.Status.Enabled);
```
Comment (rogalv) [inline src/test/java/com/emc/ecs/common/Fixtures.java:618]: I think this method needs comment on what is done with 'nonexpiration' rules, why some are disabled and others enabled :)

Comment (kirillston) [reply src/test/java/com/emc/ecs/common/Fixtures.java:638]: Added javadoc to this code

**Review by rogalv — APPROVED** [review]

**Pull request merged**